task(fxa-payments-server): add components to new checkout layout

meandavejustice opened the pull request:
## This pull request

-Adds components needed to fulfill the layout of the new checkout page.
- Adds the check to enable submit button once consent checkbox is
checked.

## Issue that this pull request solves

- fixes #9750
- refs #9361

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

![image](https://user-images.githubusercontent.com/1844554/127924286-c338b23d-37c6-4c60-a87c-ac60909fe635.png)
![image](https://user-images.githubusercontent.com/1844554/127924351-fef05b8a-7104-43aa-9504-6870a4c3689a.png)

chenba commented on packages/fxa-payments-server/src/components/Header/index.test.tsx:
```diff
@@ -33,6 +33,18 @@ describe('Header', () => {
+    const { queryByTestId } = subject();
+
+    const avatar = queryByTestId('branding');
+    expect(avatar).toBeVisible();
```
'branding' is not 'avatar'?

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -16,14 +16,18 @@ import './index.scss';
 export type NewUserEmailFormProps = {
-  getString: (id: string) => string;
+  getString?: (id: string) => string;
+  setValidEmailSet: Function;
```
Why not set the actual email?  We do need to use it in a request.  Is there any way to get the email address instead?

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -16,14 +16,18 @@ import './index.scss';
-  getString: (id: string) => string;
+  getString?: (id: string) => string;
+  setValidEmailSet: Function;
+  setAccountExists: Function;
```
If we know the types of these functions, would be good to declare them, and not just use `Function`.

meandavejustice replied:
I updated these types locally, I'm getting a typescript error in the test, I'd like to pass the correct thing here to silence the error, but I'm not sure how to fake this type, I haven't found anything terribly useful in my searches yet...

meandavejustice replied:
I actually fixed this up by just including `setState` calls in the stories and tests, it works and it's more true to the operating state of the component 👍

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -122,21 +128,26 @@ export async function checkAccountExists(userEmail: string) {
-    }
-  }
+      setAccountExists(true);
+    } else setAccountExists(false);
```
Could be
```
hasAccount = response.exists;
setAccountExists(response.exists);
```
so we won't have the nesting.

chenba commented on packages/fxa-payments-server/src/components/PaymentForm/index.tsx:
```diff
@@ -226,6 +237,67 @@ export const PaymentForm = ({
+        </SubmitButton>
+      </Localized>
+    </div>
+  ) : null;
```
When do we want `buttons` to be `null`?

meandavejustice replied:
This case is hit for when `submit` is false. In that case we would not want to show the submit button and buttons should be null.

This was left over from when I was going to provide my own submit button at the parent level. It's no longer valid or needed, I will remove.

chenba commented on packages/fxa-payments-server/src/components/PaymentForm/index.tsx:
```diff
@@ -226,6 +237,67 @@ export const PaymentForm = ({
+              id={
+                submitButtonL10nId
+                  ? submitButtonL10nId
+                  : payButtonL10nId(customer!)
```
We definitely don't have a `customer` in a couple of flows.  Instead of asserting `customer`, I think we should update `payButtonL10nId` to accept null or undefined.

meandavejustice replied:
This is an existing method, I think it would be a good thing to clean up also, can you file an issue and tag it for a maintenance sprint?

chenba replied:
I mean we are just removing a `!` here at the call site and then change the param type of `payButtonL10Id` to `(c?: Customer | null)`.  Filing an issue, triaging, creating another PR, having someone review it seems like a lot of unnecessary effort.  Would you accept a patch that you can apply? 

In this case, TypeScript is correct about the type of `customer` here.

chenba commented on packages/fxa-payments-server/src/components/PaymentForm/index.tsx:
```diff
@@ -236,7 +308,16 @@ export const PaymentForm = ({
+        <>
+          <br />
+          <br />
+          <br />
```
heh classic 😉

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.test.tsx:
```diff
@@ -104,13 +101,20 @@ describe('routes/Checkout', () => {
+
+    const planDetailsEl = getByTestId('plan-details-component');
+    expect(planDetailsEl).toBeInTheDocument();
+
```
There's also a payment form from this patch.

chenba commented:
When I opened the checkout page after building the branch locally, the layout looks different than the screenshots above.
![image](https://user-images.githubusercontent.com/198055/128035301-99bcda0f-2d64-49eb-a3a5-ebfc5eac6e99.png)

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -47,6 +99,31 @@ export const Checkout = ({
     [plans]
```
I know it's not part of this patch, but the selected plan should be based on the plan id and not the list of plans?

meandavejustice replied:
Not sure I follow, can you elaborate?

chenba replied:
That's the array of dependencies for `useMemo`, and the dependencies used in that function are `productId`, `planId`, and `plansByProductId`, not `plans`.  I think VS Code should be complaining about it to you as it does to me.

Official docs here: https://reactjs.org/docs/hooks-reference.html#usememo

meandavejustice replied:
Totally see what you mean now, I wonder why my linter wasn't picking up on that 🤔, just pushed that fix.

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -55,8 +132,138 @@ export const Checkout = ({
+        >
+          <h2 className="step-header">1. Create a Firefox account</h2>
+          <hr />
+          {/* "Already have an account? <a>Sign In</a>" https://mozilla-hub.atlassian.net/browse/FXA-3664 */}
```
Why is this commented out?

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -55,8 +132,138 @@ export const Checkout = ({
+          })}
+          data-testid="subscription-create"
+        >
+          <h2 className="step-header">1. Create a Firefox account</h2>
```
L10n?

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -55,8 +132,138 @@ export const Checkout = ({
+
+          <hr />
+          <h2 className="step-header">2. Choose your payment method</h2>
+          <h4>required</h4>
```
Is this heading markup just for the bold appearance/visual emphasis?  I don't think we should do that.

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -55,8 +132,138 @@ export const Checkout = ({
+          />
+
+          <hr />
+          <h2 className="step-header">2. Choose your payment method</h2>
```
L10n?

chenba reviewed (changes requested):
Thanks for the big patch, Dave.  Got some questions and suggestions in the comments.  Also, the page in my local env doesn't seem to match your screenshots (fwiw I have PayPal enabled).

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -16,14 +16,18 @@ import './index.scss';
-  checkAccountExists: Function;
+  getString?: (id: string) => string;
+  setValidEmail: Dispatch<SetStateAction<string>>;
+  setAccountExists: Dispatch<SetStateAction<boolean>>;
```
nit: Those are definitely the types for those functions.  But they are harder for other engineers to read.  `(value: string) => void;` and `(value: boolean) => void;` make it obvious what the param and return types are.

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -41,9 +45,9 @@ export const NewUserEmailForm = ({
+      <Localized id="new-user-sign-in-link" elems={{ a: <a></a> }}>
         <p className="sign-in-copy" data-testid="sign-in-copy">
-          Already have a Firefox account? <a>Sign in</a>
+          Already have a Firefox account? <a href="#">Sign in</a>
```
Any reason we don't have the actual url for the href here?

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -122,21 +128,24 @@ export async function checkAccountExists(userEmail: string) {
   setEmailInputState: Function,
-  getString: Function,
-  checkAccountExists: Function
+  checkAccountExists: Function,
```
I see that you defined the type at https://github.com/mozilla/fxa/pull/10009/files#diff-e7d37441df07a444329b2215c48e3c74a7a1c5d26f586cea6958da8bb385a9a8R22 Why not use it here?  `setValidEmail` and `setAccountExists` too.

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx:
```diff
@@ -122,21 +128,24 @@ export async function checkAccountExists(userEmail: string) {
-  }
+    hasAccount = response.exists;
+    setAccountExists(response.exists);
+  } else setValidEmail('');
```
nit: even though there's one line for the `else`, have it in a `{}` block is still preferred.

chenba reviewed (changes requested):
@meandavejustice  thanks for the updated patch.  I got some nits, but more importantly:

- We should use the actual url for the href of the link that takes the user with an account to sign in; afaik there isn't another jira for doing that
- I actually cannot click on the "sign in" link above the text input; something might be blocking it?
- Does the PayPal button display for you?  It doesn't for me.

chenba commented:
Is the checkbox above the PayPal button suppose to affect the state of the PayPal button?  Currently I can click on the PayPal button without checking the checkbox.

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.stories.tsx:
```diff
@@ -1,21 +1,35 @@
     <div style={{ display: 'flex' }}>
       <NewUserEmailForm
+        signInURL={
+          'https://localhost:3031/subscriptions/products/productId?plan=planId&signin=yes'
```
Hard coded to `localhost` because we are changing it in the other jira?

chenba replied:
oh haha this is for storybook.  my bad 🤦

chenba commented:
> 
> 
> Is the checkbox above the PayPal button suppose to affect the state of the PayPal button? Currently I can click on the PayPal button without checking the checkbox.

Oh kind of answering my question... I noticed that #9361 is listed as one of the issues that this PR fixes.

chenba commented on packages/fxa-payments-server/src/components/NewUserEmailForm/index.test.tsx:
```diff
@@ -136,6 +142,7 @@ describe('NewUserEmailForm test', () => {
       'foxy@mozilla.com',
       false,
       (state: string) => state,
+      (id: boolean) => id,
```
`id`?

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.scss:
```diff
@@ -1,3 +1,103 @@
+
+  .subscription-create-pay-with-other .paypal-button {
+    width: 240px;
+    height: 48px + 48;
```
Is this the same as having `96px`?

meandavejustice replied:
yeah not sure why it was written this way, copied from `subscriptionCreate/index.scss`. happy to change it

chenba commented on packages/fxa-payments-server/src/routes/Checkout/index.tsx:
```diff
@@ -55,8 +129,137 @@ export const Checkout = ({
+                      <PaypalButton
+                        currencyCode={selectedPlan.currency}
+                        customer={null}
+                        idempotencyKey={'abc'}
```
Is there a follow-up jira for this?

chenba reviewed (changes requested):
@meandavejustice it's looking pretty good.  That checkbox doesn't seem have any effect on the PayPal button though.

meandavejustice commented:
@chenba 
> Oh kind of answering my question... I noticed that #9361 is listed as one of the issues that this PR fixes.

I originally mistook this issue for only disabling the submit button. I'll look at disabling the paypal button also as a part of this

meandavejustice commented:
@chenba Addressed comments and updated description to read as "refs #9361" so I can complete the paypal disabling without holding this layout pr any longer.

The paypal disable needs to be called from within the `onInit` function passed into the button. The only issue is how we are to call this method again, once the `disabled` prop is updated. I'd like to think about this a tad more to make sure we doing it the React ™️ way.

chenba commented:
> 
> 
> @chenba Addressed comments and updated description to read as "refs #9361"...

Cool, can you also update the commit message?

chenba reviewed (approved):
💰
